agent/src/main/scala/za/co/absa/atum/agent/AtumContext.scala

```diff
@@ -0,0 +1,77 @@
+package za.co.absa.atum.agent
+
+import org.apache.spark.sql.DataFrame
+import za.co.absa.atum.agent.model.{MeasureResult, Measurement}
+
+/**
+ *  AtumContext: This class provides the methods to measure Spark `Dataframe`. Also allows to add/edit/remove measures.
+ *  @param measurements: A sequences of measurements.
+ */
+
+case class AtumContext(measurements: Set[Measurement] = Set()) {
+
+  def withMeasures(
+    measure: Measurement
+  ): AtumContext =
+    this.copy(measurements = Set(measure))
+
+  def withMeasures(
+    measures: Iterable[Measurement]
+  ): AtumContext =
+    this.copy(measurements = measures.toSet)
+
+  def withMeasuresAdded(
+    measure: Measurement
+  ): AtumContext =
+    this.copy(measurements = measurements + measure)
+
+  def withMeasuresAdded(
+    measures: Iterable[Measurement]
+  ): AtumContext =
+    this.copy(measurements = measurements ++ measures)
+
+  def withMeasureRemoved(measurement: Measurement): AtumContext =
+    this.copy(measurements = measurements.filterNot(_ == measurement))
+
+}
+
+object AtumContext {
+  implicit class DatasetWrapper(df: DataFrame) {
+
+    /**
+     *  Executes the measure directly with not AtumContext.
+     *  @param measure the measure to be calculated
+     *  @return
+     */
+    def executeMeasure(checkpointName: String, measure: Measurement): DataFrame = {
+
+      val result = MeasureResult(measure, measure.function(df))
+      AtumAgent.measurePublish(checkpointName, result)
```

Comment:
actually, what's the difference between publish and measurePublish?

Reply:
`publish` is the main case. You submit an `AtumContext` and the method sends the data to the API. `measurePublish` is a corner case where you only submit a bunch of `Measurement`.

`AtumContext` is going to provide metadata with `Segmentation` in the next iteration of code so it makes sense to have an AtumContext